// @editor/quick-compiler/dist/file-uid.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeFilePath = exports.encodeRegularFilePath = exports.decodeRegularFilePath = exports.decodeVirtualFileUid = exports.encodeVirtualFileUid = exports.isVirtualFileUid = void 0;
const path_1 = __importDefault(require("path"));
function isVirtualFileUid(fileId) {
    return fileId.codePointAt(0) === 0;
}
exports.isVirtualFileUid = isVirtualFileUid;
function encodeVirtualFileUid(moduleId) {
    return `\0${moduleId}`;
}
exports.encodeVirtualFileUid = encodeVirtualFileUid;
function decodeVirtualFileUid(fileId) {
    return fileId.substr(1);
}
exports.decodeVirtualFileUid = decodeVirtualFileUid;
function decodeRegularFilePath(fileUid) {
    return fileUid;
}
exports.decodeRegularFilePath = decodeRegularFilePath;
function encodeRegularFilePath(path) {
    return path;
}
exports.encodeRegularFilePath = encodeRegularFilePath;
function encodeFilePath(path, rootDir) {
    const relativeFromRoot = path_1.default.relative(path_1.default.join(rootDir), path);
    if (relativeFromRoot.length === 0 ||
        relativeFromRoot.startsWith('..') ||
        path_1.default.isAbsolute(relativeFromRoot)) {
        return null;
    }
    else {
        return relativeFromRoot;
    }
}
exports.encodeFilePath = encodeFilePath;
