// @editor/quick-compiler/dist/indexed-source-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexedSourceMap = void 0;
class IndexedSourceMap {
    constructor({ file, separator, }) {
        this._sourceMap = {
            version: 3,
            sections: [],
        };
        this._code = '';
        this._lines = 0;
        this._columns = 0;
        this._separator = '';
        this._sourceMap = {
            version: 3,
            file,
            sections: [],
        };
        this._separator = separator !== null && separator !== void 0 ? separator : '';
    }
    get code() {
        return this._code;
    }
    get sourceMap() {
        return JSON.stringify(this._sourceMap);
    }
    add(code, map) {
        if (Buffer.isBuffer(code)) {
            code = code.toString();
        }
        if (this._separator) {
            this._addCode(this._separator);
        }
        if (map) {
            this._sourceMap.sections.push({
                offset: { line: this._lines, column: this._columns },
                map,
            });
        }
        this._addCode(code);
    }
    _addCode(code) {
        const lines = code.split(/\r\n|\r|\n/g);
        this._lines += lines.length - 1;
        this._columns += lines[lines.length - 1].length;
        this._code += code;
    }
}
exports.IndexedSourceMap = IndexedSourceMap;
