// @cocos/ccbuild/modules/modularize/lib/platform-config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebPlatform = exports.NativePlatform = exports.MinigamePlatform = void 0;
var MinigamePlatform;
(function (MinigamePlatform) {
  MinigamePlatform[MinigamePlatform["WECHAT"] = 0] = "WECHAT";
  MinigamePlatform[MinigamePlatform["WECHAT_MINI_PROGRAM"] = 1] = "WECHAT_MINI_PROGRAM";
  /**
   * @deprecated this platform has been removed.
   */
  MinigamePlatform[MinigamePlatform["BAIDU"] = 2] = "BAIDU";
  MinigamePlatform[MinigamePlatform["BYTEDANCE"] = 3] = "BYTEDANCE";
  MinigamePlatform[MinigamePlatform["XIAOMI"] = 4] = "XIAOMI";
  MinigamePlatform[MinigamePlatform["ALIPAY"] = 5] = "ALIPAY";
  MinigamePlatform[MinigamePlatform["TAOBAO"] = 6] = "TAOBAO";
  MinigamePlatform[MinigamePlatform["TAOBAO_MINIGAME"] = 7] = "TAOBAO_MINIGAME";
  MinigamePlatform[MinigamePlatform["OPPO"] = 8] = "OPPO";
  MinigamePlatform[MinigamePlatform["VIVO"] = 9] = "VIVO";
  MinigamePlatform[MinigamePlatform["HUAWEI"] = 10] = "HUAWEI";
  /**
   * @deprecated this platform has been removed.
   */
  MinigamePlatform[MinigamePlatform["COCOSPLAY"] = 11] = "COCOSPLAY";
  /**
   * @deprecated this platform has been removed.
   */
  MinigamePlatform[MinigamePlatform["QTT"] = 12] = "QTT";
  /**
   * @deprecated this platform has been removed.
   */
  MinigamePlatform[MinigamePlatform["LINKSURE"] = 13] = "LINKSURE";
})(MinigamePlatform = exports.MinigamePlatform || (exports.MinigamePlatform = {}));
var NativePlatform;
(function (NativePlatform) {
  NativePlatform[NativePlatform["NATIVE_EDITOR"] = 0] = "NATIVE_EDITOR";
  NativePlatform[NativePlatform["ANDROID"] = 1] = "ANDROID";
  NativePlatform[NativePlatform["WINDOWS"] = 2] = "WINDOWS";
  NativePlatform[NativePlatform["IOS"] = 3] = "IOS";
  NativePlatform[NativePlatform["MAC"] = 4] = "MAC";
  NativePlatform[NativePlatform["OHOS"] = 5] = "OHOS";
  NativePlatform[NativePlatform["OPEN_HARMONY"] = 6] = "OPEN_HARMONY";
  NativePlatform[NativePlatform["LINUX"] = 7] = "LINUX";
})(NativePlatform = exports.NativePlatform || (exports.NativePlatform = {}));
var WebPlatform;
(function (WebPlatform) {
  WebPlatform[WebPlatform["WEB_EDITOR"] = 0] = "WEB_EDITOR";
  WebPlatform[WebPlatform["WEB_MOBILE"] = 1] = "WEB_MOBILE";
  WebPlatform[WebPlatform["WEB_DESKTOP"] = 2] = "WEB_DESKTOP";
})(WebPlatform = exports.WebPlatform || (exports.WebPlatform = {}));
