// @editor/creator/static/theme/creator.js
'use strict';

exports.version = 1;

exports.font = {
    normal: `BlinkMacSystemFont,'Helvetica Neue',Helvetica,'Lucida Grande','Segoe UI',Ubuntu,Cantarell,SourceHanSansCN-Normal,Arial,sans-serif`,
    editor: `Menlo,Monaco,"Courier New",monospace`,
};

exports.contrast = 0.6;

exports.color = {
    // 正常页面配色
    normal: {
        // 背景颜色
        fill: {
            important: '#1F1F1F',
            emphasis: '#141414',
            normal: '#2b2b2b',
            weaker: '#3D3D3D',
            weakest: '#525252',
        },
        // 字体颜色
        contrast: {
            important: '#CCCCCC',
            emphasis: '#8F8F8F',
            normal: '#CCCCCC',
            weaker: '#CCCCCC',
            weakest: '#CCCCCC',
        },
        // 边框颜色
        border: {
            important: '#050505',
            emphasis: '#050505',
            normal: '#050505',
            weaker: '#050505',
            weakest: '#050505',
        },
    },

    // 物体聚焦的时候的配色
    focus: {
        fill: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
        contrast: {
            important: '#CCCCCC',
            emphasis: '#fafafa',
            normal: '#CCCCCC',
            weaker: '#CCCCCC',
            weakest: '#CCCCCC',
        },
        border: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
    },

    // 鼠标移入物体的配色
    hover: {
        fill: {
            important: '#1F1F1F',
            emphasis: '#2b2b2b',
            normal: '#3D3D3D',
            weaker: '#525252',
            weakest: '#8F8F8F',
        },
        contrast: {
            important: '#CCCCCC',
            emphasis: '#CCCCCC',
            normal: '#CCCCCC',
            weaker: '#CCCCCC',
            weakest: '#CCCCCC',
        },
        border: {
            important: '#050505',
            emphasis: '#227F9B',
            normal: '#227F9B',
            weaker: '#050505',
            weakest: '#050505',
        },
    },

    // 物体激活状态的时候的配色
    active: {
        fill: {
            important: '#2b2b2b',
            emphasis: '#3D3D3D',
            normal: '#525252',
            weaker: '#8F8F8F',
            weakest: '#A3A3A3',
        },
        contrast: {
            important: '#CCCCCC',
            emphasis: '#CCCCCC',
            normal: '#CCCCCC',
            weaker: '#CCCCCC',
            weakest: '#CCCCCC',
        },
        border: {
            important: '#050505',
            emphasis: '#3D3D3D',
            normal: '#525252',
            weaker: '#8F8F8F',
            weakest: '#A3A3A3',
        },
    },

    // 重要的元素
    primary: {
        fill: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
        contrast: {
            important: '#FAFAFA',
            emphasis: '#FAFAFA',
            normal: '#FAFAFA',
            weaker: '#FAFAFA',
            weakest: '#000000',
        },
        border: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
    },
    // 图标按钮
    icon: {
        fill: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
        contrast: {
            important: '#FAFAFA',
            emphasis: '#FAFAFA',
            normal: '#FAFAFA',
            weaker: '#FAFAFA',
            weakest: '#000000',
        },
        border: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
    },

    // 默认元素
    default: {
        fill: {
            important: '#3D3D3D',
            emphasis: '#2b2b2b',
            normal: '#3d3d3d',
            weaker: '#525252',
            weakest: '#707070',
        },
        contrast: {
            important: '#cccccc',
            emphasis: '#A3A3A3',
            normal: '#cccccc',
            weaker: '#cccccc',
            weakest: '#cccccc',
        },
        border: {
            important: '#3D3D3D',
            emphasis: '#2b2b2b',
            normal: '#3d3d3d',
            weaker: '#424242',
            weakest: '#707070',
        },
    },

    // 显示成功消息的元素
    success: {
        fill: {
            important: '#198F6B',
            emphasis: '#32B086',
            normal: '#64DAAF',
            weaker: '#8CE3BD',
        },
        contrast: {
            important: '#fafafa',
            emphasis: '#fafafa',
            normal: '#fafafa',
            weaker: '#fafafa',
            weakest: '#fafafa',
        },
        border: {
            important: '#198F6B',
            emphasis: '#32B086',
            normal: '#64DAAF',
            weaker: '#8CE3BD',
        },
    },

    // 普通信息元素
    info: {
        fill: {
            important: '#094A5D',
            emphasis: '#135C73',
            normal: '#227F9B',
            weaker: '#40AACA',
            weakest: '#65BDD7',
        },
        contrast: {
            important: '#FAFAFA',
            emphasis: '#FAFAFA',
            normal: '#FAFAFA',
            weaker: '#FAFAFA',
            weakest: '#FAFAFA',
        },
        border: {
            important: '#135C73',
            emphasis: '#227F9B',
            normal: '#40AACA',
            weaker: '#65BDD7',
            weakest: '#AFDFEE',
        },
    },

    // 危险元素
    danger: {
        fill: {
            important: '#8D1833',
            emphasis: '#AE2D47',
            normal: '#D85F6F',
            weaker: '#E27F88',
        },
        contrast: {
            important: '#fafafa',
            emphasis: '#fafafa',
            normal: '#fafafa',
            weaker: '#fafafa',
        },
        border: {
            important: '#8D1833',
            emphasis: '#AE2D47',
            normal: '#D85F6F',
            weaker: '#E27F88',
        },
    },

    // 警告元素
    warn: {
        fill: {
            important: '#753307',
            emphasis: '#C66F23',
            normal: '#F1A348',
            weaker: '#F5BB70',
        },
        contrast: {
            important: '#fafafa',
            emphasis: '#fafafa',
            normal: '#fafafa',
            weaker: '#fafafa',
        },
        border: {
            important: '#753307',
            emphasis: '#C66F23',
            normal: '#F1A348',
            weaker: '#F5BB70',
        },
    },

    selection: {
        fill: {
            normal: '#0966d5',
        },
        contrast: {
            normal: '#ffffff',
        },
    }
};

exports.size = {
    // 普通
    normal: {
        line: 20,    // 行高
        font: 12,    // 文字大小
        radius: 2,   // 边角半径
        border: 1,   // 边框
    },

    // 超大
    large: {
        line: 32,
        font: 16,
        radius: 2,
        border: 1,
    },

    // 偏大
    big: {
        line: 28,
        font: 14,
        radius: 2,
        border: 1,
    },

    // 小
    small: {
        line: 20,
        font: 10,
        radius: 2,
        border: 1,
    },

    // 很小
    little: {
        line: 16,
        font: 8,
        radius: 2,
        border: 1,
    },
};
