// @cocos/ccbuild/modules/bundler/lib/rollup/plugins.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  var desc = Object.getOwnPropertyDescriptor(m, k);
  if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
    desc = {
      enumerable: true,
      get: function () {
        return m[k];
      }
    };
  }
  Object.defineProperty(o, k2, desc);
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});
var __setModuleDefault = this && this.__setModuleDefault || (Object.create ? function (o, v) {
  Object.defineProperty(o, "default", {
    enumerable: true,
    value: v
  });
} : function (o, v) {
  o["default"] = v;
});
var __importStar = this && this.__importStar || function (mod) {
  if (mod && mod.__esModule) return mod;
  var result = {};
  if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
  __setModuleDefault(result, mod);
  return result;
};
var __importDefault = this && this.__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    "default": mod
  };
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.typescript = exports.terser = exports.virtual = exports.commonjs = exports.nodeResolve = exports.json = exports.babel = void 0;
const babel = __importStar(require("@rollup/plugin-babel"));
exports.babel = babel;
const plugin_json_1 = __importDefault(require("@rollup/plugin-json"));
exports.json = plugin_json_1.default;
const plugin_node_resolve_1 = __importDefault(require("@rollup/plugin-node-resolve"));
exports.nodeResolve = plugin_node_resolve_1.default;
const plugin_commonjs_1 = __importDefault(require("@rollup/plugin-commonjs"));
exports.commonjs = plugin_commonjs_1.default;
const plugin_virtual_1 = __importDefault(require("@rollup/plugin-virtual"));
exports.virtual = plugin_virtual_1.default;
const rollup_plugin_terser_1 = __importDefault(require("rollup-plugin-terser"));
exports.terser = rollup_plugin_terser_1.default;
const plugin_typescript_1 = __importDefault(require("@rollup/plugin-typescript"));
exports.typescript = plugin_typescript_1.default;
