// editor/index.js
'use strict';

exports.App = require('@editor/creator/dist/app').App;
exports.Dialog = require('@editor/creator/dist/dialog').Dialog;
exports.EditMode = require('@editor/creator/dist/edit-mode').EditMode;
exports.I18n = require('@editor/creator/dist/i18n').I18n;
exports.Message = require('@editor/creator/dist/message').Message;
exports.Layout = require('@editor/creator/dist/layout').Layout;
// exports.Logger = require('@editor/creator/dist/logger').Logger;
exports.Menu = require('@editor/creator/dist/menu').Menu;
exports.Metrics = require('@editor/creator/dist/metrics').Metrics;
exports.Network = require('@editor/creator/dist/network').Network;
exports.Package = require('@editor/creator/dist/package').Package;
exports.Panel = require('@editor/creator/dist/panel').Panel;
exports.Profile = require('@editor/creator/dist/profile').Profile;
exports.Project = require('@editor/creator/dist/project').Project;
exports.Selection = require('@editor/creator/dist/selection').Selection;
// exports.Startup = require('@editor/creator/dist/startup').Startup;
exports.Task = require('@editor/creator/dist/task').Task;
exports.Theme = require('@editor/creator/dist/theme').Theme;
exports.UI = require('@editor/creator/dist/ui-kit').UI;
exports.User = require('@editor/creator/dist/user').User;
exports.Utils = require('@editor/creator/dist/utils').Utils;
exports.Module = require('@editor/creator/dist/module').Module;
exports.Clipboard = require('@editor/creator/dist/clipboard').Clipboard;

const { join } = require('path');

// worker 进程
if (global && global.window && global.window.__name__) {
    process.nextTick(() => {
        // Hack 监听 message 数据
        const mWindows = require(join(exports.App.path, './builtin/messages/static/windows.js'));
        mWindows.load();
    });
}
