// @editor/lib-programming/dist/build-systemjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = void 0;
const rollup = __importStar(require("rollup"));
const rollup_plugin_terser_1 = require("rollup-plugin-terser");
const plugin_replace_1 = __importDefault(require("@rollup/plugin-replace"));
const resolve_systemjs_specifier_1 = __importDefault(require("./resolve-systemjs-specifier"));
const virtual_entry_source_1 = __importDefault(require("./virtual-entry-source"));
async function build({ input, inputSource, sourceMap, minify, out, browser, format, }) {
    const modules = {};
    if (inputSource !== undefined) {
        modules[input] = inputSource;
    }
    await (await rollup.rollup({
        input,
        plugins: [
            virtual_entry_source_1.default(modules),
            resolve_systemjs_specifier_1.default,
            plugin_replace_1.default({
                // @ts-ignore This option in `@rollup/plugin-replace@2.3.3` is default to `false`.
                // It is recommended to set this option to `true`, as the next major version will default this option to `true`.
                preventAssignment: false,
                'process.env.SYSTEM_PRODUCTION': minify ? 'true' : 'false',
                'process.env.SYSTEM_BROWSER': (browser !== null && browser !== void 0 ? browser : true) ? 'true' : 'false',
            }),
            minify ? rollup_plugin_terser_1.terser({}) : undefined,
        ],
    })).write({
        file: out,
        sourcemap: sourceMap,
        format: format !== null && format !== void 0 ? format : 'iife',
    });
}
exports.build = build;
