// @editor/extension-sdk/sdk/manager/download/download.constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnableDownloadStatus = exports.EDownloadStatus = void 0;
var EDownloadStatus;
(function (EDownloadStatus) {
    EDownloadStatus[EDownloadStatus["PREPARING"] = 0] = "PREPARING";
    EDownloadStatus[EDownloadStatus["READY"] = 1] = "READY";
    EDownloadStatus[EDownloadStatus["LOADING"] = 2] = "LOADING";
    EDownloadStatus[EDownloadStatus["LOADED"] = 3] = "LOADED";
    EDownloadStatus[EDownloadStatus["COMPRESSING"] = 4] = "COMPRESSING";
    EDownloadStatus[EDownloadStatus["INSTALLING"] = 5] = "INSTALLING";
    EDownloadStatus[EDownloadStatus["COMPLETE"] = 6] = "COMPLETE";
    EDownloadStatus[EDownloadStatus["CANCELED"] = 7] = "CANCELED";
    EDownloadStatus[EDownloadStatus["FAILED"] = 8] = "FAILED";
})(EDownloadStatus = exports.EDownloadStatus || (exports.EDownloadStatus = {}));
exports.EnableDownloadStatus = [
    EDownloadStatus.PREPARING,
    EDownloadStatus.CANCELED,
    EDownloadStatus.FAILED,
    EDownloadStatus.COMPLETE,
];
