// @cocos/creator-programming-babel-preset-cc/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const plugin_transform_typescript_1 = __importDefault(require("@babel/plugin-transform-typescript"));
// @ts-ignore
const plugin_proposal_decorators_1 = __importDefault(require("@babel/plugin-proposal-decorators"));
// @ts-ignore
const helper_plugin_utils_1 = require("@babel/helper-plugin-utils");
// @ts-ignore
const babel_plugin_const_enum_1 = __importDefault(require("babel-plugin-const-enum"));
const $ = (0, helper_plugin_utils_1.declare)((api, { allowDeclareFields, onlyRemoveTypeImports, useDefineForClassFields, }) => {
    api.assertVersion(7);
    const looseClassProperties = !useDefineForClassFields;
    // Note, if we choose to use @babel/preset-typescript instead of @babel/plugin-transform-typescript
    // We have to wrap the `plugins` into an anonymous preset and that preset should
    // be placed prior to @babel/preset-typescript.
    // Otherwise, the @babel/proposal-class-properties would execute before @babel/plugin-transform-typescript.
    // The result would be, unexpectly, even if we pass `loose: true` && `allowDeclareFields: false`
    // to @babel/preset-typescript, class fields with no explicit initializer was still not intent to be erased.
    // This behaviour does not match what tsc does as tsconfig `useDefineForClassFields` enabled.
    return {
        overrides: [{
                plugins: [
                    // TS const enum
                    [babel_plugin_const_enum_1.default, {
                            transform: 'constObject'
                        }],
                    [plugin_transform_typescript_1.default, {
                            allowNamespaces: true,
                            allowDeclareFields,
                            onlyRemoveTypeImports,
                        }],
                    [plugin_proposal_decorators_1.default, {
                            legacy: true
                        }],
                ]
            }],
    };
});
exports.default = $;
