// @cocos/creator-programming-mod-lo/lib/transformer/babel/babel-plugins/plugin-check-obsolete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginCheckObsolete = void 0;
const core_1 = require("@babel/core");
const templateOption = {
    syntacticPlaceholders: true,
};
exports.pluginCheckObsolete = {
    visitor: {
        Program(path) {
            let needImportHelper = false;
            path.traverse({
                ImportDeclaration(path) {
                    if (path.node.source.value !== 'cc') {
                        return;
                    }
                    needImportHelper = true;
                    if (path.node.specifiers[0].type === 'ImportNamespaceSpecifier') {
                        // handle namespace import
                        const importSpecifier = path.node.specifiers[0];
                        const localName = importSpecifier.local.name;
                        const newLocalName = path.scope.generateUid(localName);
                        path.replaceWith(core_1.template.statement(`import * as ${newLocalName} from 'cc';`, templateOption)());
                        path.insertAfter(core_1.template.statement(`const ${localName} = __checkObsoleteInNamespace__(${newLocalName});`, templateOption)());
                        path.skip();
                    }
                    else {
                        const imported = [];
                        path.traverse({
                            ImportSpecifier(path) {
                                imported.push(`'${path.node.imported.name}'`);
                            },
                        });
                        if (imported.length > 0) {
                            const checkObsoleteStatement = core_1.template.statement(`__checkObsolete__([${imported.join(', ')}]);`, templateOption)();
                            path.insertAfter(checkObsoleteStatement);
                        }
                    }
                }
            });
            // insert import helper statement
            if (needImportHelper) {
                const importCheckObsolete = core_1.template.statement(`import { __checkObsolete__, __checkObsoleteInNamespace__ } from 'cc';`, templateOption)();
                path.node.body = [importCheckObsolete].concat(path.node.body);
            }
        },
    }
};
