// @editor/creator/.build.config.js
const { join } = require('path');

exports.tsc = function() {
    return [
        './',
    ];
};

exports.less = function() {
    const list = [
        // dev
        {
            source: join(__dirname, './source/theme/source/dock.less'),
            dist: join(__dirname, './dev/theme/dist/dock.css'),
        },
        {
            source: join(__dirname, './source/theme/source/panel.less'),
            dist: join(__dirname, './dev/theme/dist/panel.css'),
        },
        // dist
        {
            source: join(__dirname, './source/theme/source/dock.less'),
            dist: join(__dirname, './dist/theme/dist/dock.css'),
        },
        {
            source: join(__dirname, './source/theme/source/panel.less'),
            dist: join(__dirname, './dist/theme/dist/panel.css'),
        },
    ];
    return list;
};

exports.compression = function() {
    return {
        source: join(__dirname, './dev'),
        target: join(__dirname, './dist'),
    };
};

exports.encryption = function() {
    return {
        source: join(__dirname, './dist'),
        exclude: [
            './dist/index.js',
            './dist/require.js',
        ],
    };
};
