// @cocos/ccbuild/lib/src/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  var desc = Object.getOwnPropertyDescriptor(m, k);
  if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
    desc = {
      enumerable: true,
      get: function () {
        return m[k];
      }
    };
  }
  Object.defineProperty(o, k2, desc);
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});
var __setModuleDefault = this && this.__setModuleDefault || (Object.create ? function (o, v) {
  Object.defineProperty(o, "default", {
    enumerable: true,
    value: v
  });
} : function (o, v) {
  o["default"] = v;
});
var __importStar = this && this.__importStar || function (mod) {
  if (mod && mod.__esModule) return mod;
  var result = {};
  if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
  __setModuleDefault(result, mod);
  return result;
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dtsBundler = exports.Modularize = exports.Bundler = exports.Transformer = exports.ConfigInterface = exports.StatsQuery = exports.buildEngine = void 0;
const Transformer = __importStar(require("../../modules/transformer/lib/index.js"));
exports.Transformer = Transformer;
const Bundler = __importStar(require("../../modules/bundler/lib/index.js"));
exports.Bundler = Bundler;
const Modularize = __importStar(require("../../modules/modularize/lib/index.js"));
exports.Modularize = Modularize;
const dtsBundler = __importStar(require("../../modules/dts-bundler/lib/index.js"));
exports.dtsBundler = dtsBundler;
var build_engine_1 = require("../../modules/build-engine/lib/index.js");
Object.defineProperty(exports, "buildEngine", {
  enumerable: true,
  get: function () {
    return build_engine_1.buildEngine;
  }
});
var stats_query_1 = require("../../modules/stats-query/lib/index.js");
Object.defineProperty(exports, "StatsQuery", {
  enumerable: true,
  get: function () {
    return stats_query_1.StatsQuery;
  }
});
Object.defineProperty(exports, "ConfigInterface", {
  enumerable: true,
  get: function () {
    return stats_query_1.ConfigInterface;
  }
});
